internal/topology/variables/cluster_variable_validation.go

```diff
@@ -0,0 +1,124 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package variables
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"k8s.io/apiextensions-apiserver/pkg/apis/apiextensions"
+	"k8s.io/apiextensions-apiserver/pkg/apiserver/validation"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+)
+
+// ValidateClusterVariables validates clusterVariable via the schemas in the corresponding clusterClassVariable.
+func ValidateClusterVariables(clusterVariables []clusterv1.ClusterVariable, clusterClassVariables []clusterv1.ClusterClassVariable, fldPath *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+
+	// Build maps for easier and faster access.
+	clusterVariablesMap := map[string]*clusterv1.ClusterVariable{}
+	for i := range clusterVariables {
+		clusterVariablesMap[clusterVariables[i].Name] = &clusterVariables[i]
+	}
+	clusterClassVariablesMap := map[string]*clusterv1.ClusterClassVariable{}
+	for i := range clusterClassVariables {
+		clusterClassVariablesMap[clusterClassVariables[i].Name] = &clusterClassVariables[i]
+	}
+
+	// Validate:
+	// * required variables from the ClusterClass exist on the Cluster.
+	// * all variables in the Cluster are defined in the ClusterClass.
+	allErrs = append(allErrs, validateRequiredClusterVariables(clusterVariablesMap, clusterClassVariablesMap, fldPath)...)
+	allErrs = append(allErrs, validateClusterVariablesDefined(clusterVariablesMap, clusterClassVariablesMap, fldPath)...)
+	if len(allErrs) > 0 {
+		return allErrs
+	}
+
+	// Validate all variables from the Cluster.
+	for variableName, clusterVariable := range clusterVariablesMap {
+		// Get schema.
+		// Nb. We already validated above in validateClusterVariablesDefined that there is a
+		// corresponding ClusterClass variable, so we don't have to check it again.
+		clusterClassVariable := clusterClassVariablesMap[variableName]
+
+		allErrs = append(allErrs, validateClusterVariable(clusterVariable, clusterClassVariable, fldPath)...)
```

Comment:
Q: can we merge `validateRequiredClusterVariables` and `validateClusterVariablesDefined` into  `validateClusterVariable`

Reply:
I would prefer not doing it as then we have to pass the full maps of variables into `validateClusterVariable`. Essentially we would move l.44-l.59 into `validateClusterVariable`. The idea of `validateClusterVariable` is that it currently validates a **single** variable which is already not trivial. When we move l.44-l.59 into `validateClusterVariable`, `validateClusterVariable` essentially becomes `ValidateClusterVariables` (+/- the map pre-calculation)

Additional context: the way `validateClusterVariable` is scoped today we can also reuse it later to validate a list of variable values against **1** variable schema, which should be useful to efficiently validate variable schema changes in a ClusterClass against a list of Clusters. (because we have to instantiate the schema validator structs only once per variable schema)

Reply:
I think it's also good to have the simple pattern of having a func for ClusterClassVariable validation, ClusterVariable validation, ClusterVariable defaulting which works on a single instance. This should help us scale up the logic if we introduce: object variables, array variables, defaulting on object/array variables, potentially advanced logic when a variable is required, ..